backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ChatMessageEventHandler.java

```diff
@@ -0,0 +1,107 @@
+import org.hyperledger.bpa.model.ChatMessage;
+import org.hyperledger.bpa.repository.PartnerRepository;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.UUID;
+
+@Slf4j
+@NoArgsConstructor
+@Singleton
+public class ChatMessageEventHandler {
+
+    @Inject
+    PartnerRepository partnerRepo;
+
+    @Inject
+    ConnectionManager cm;
+
+    @Inject
+    ChatMessageManager chatMessageManager;
+
+    @Inject
+    NotificationManager notificationManager;
+
+    public void handleIncomingMessage(@NonNull BasicMessage basicMessage) {
+        // message coming in from aries...
+        // workflow:
+        // check for content
+        // check for partner
+        // all ok, then save it
+        // if saved, send notification to Frontend for new message
+        if (StringUtils.isNotEmpty(basicMessage.getContent())) {
+            // find the partner...
+            partnerRepo.findByConnectionId(basicMessage.getConnectionId()).ifPresentOrElse(p -> {
+                // great, we have a partner...
+                // save the message
+                ChatMessage chatMessage = chatMessageManager.saveIncomingMessage(p, basicMessage.getContent());
+                if (chatMessage != null && chatMessage.getId() != null) {
+                    // notify frontend
+                    notificationManager.newIncomingMessage(p, chatMessage);
+                } else {
+                    log.error(
+                            "Error handling incoming basic message. Chat Message not saved. Partner ID = '%s', Connection ID = '%s', Message ID = '%s'",
+                            p.getId().toString(), basicMessage.getConnectionId(), basicMessage.getMessageId());
+                }
+            },
+                    () -> {
+                        log.error("Error handling incoming basic message. Partner not found for Connection ID = '%s'",
+                                basicMessage.getConnectionId());
+                    });
+        } else {
+            log.debug("Not persisting basic message. Message has no content. Connection ID = '%s', Message ID = '%s'",
+                    basicMessage.getConnectionId(), basicMessage.getMessageId());
+        }
+    }
+
+    public void handleOutgoingMessage(@NonNull String partnerId, @NonNull String content) {
+        // want to send an outgoing message to a partner...
+        // workflow:
+        // get partner, check if we can send (connection id and aries support required)
+        // all ok, then send it via ARIES
+        // if send, then we need to persist it
+        partnerRepo.findById(UUID.fromString(partnerId)).ifPresentOrElse(p -> {
+            if (StringUtils.isNotEmpty(p.getConnectionId()) && p.getAriesSupport()) {
+                if (cm.sendMessage(p.getConnectionId(), content)) {
+                    ChatMessage chatMessage = chatMessageManager.saveOutgoingMessage(p, content);
+                    if (chatMessage == null || chatMessage.getId() == null) {
```

Comment:
I never tested that but I would assume that a model is either persisted or an exception is thrown.

Reply:
It should throw an exception from Micronaut.
Changing to raise and have it auto-handled at the controller level.